numpy/polynomial/polyutils.py

```diff
@@ -725,6 +725,179 @@ def _fit(vander_f, x, y, deg, rcond=None, full=False, w=None):
+        raise ValueError("Need to set some axis")
+
+    # Flatten input
+    coords = list(coords)
+    for i in range(ndim):
+        coords[i] = np.asarray(coords[i]) + 0.0
+    data = np.asarray(data) + 0.0
+
+    deg = np.asarray(deg)
+    if w is not None:
+        w = np.asarray(w) + 0.0
+
+    if deg.dtype.kind not in 'iu' or deg.size == 0:
+        raise TypeError("deg must be an int or non-empty array of int")
+    if deg.ndim == 1 and not deg.size == ndim:
+        raise ValueError(f"deg must be of length {ndim}, if it is a 1d array")
+    if deg.ndim == ndim and np.all(deg == 0):
+        raise ValueError("deg must have at least one non-zero value")
+    if deg.ndim not in [0, 1, ndim]:
+        raise TypeError(f"deg must be an array of dimension {ndim}")
+    if deg.min() < 0:
+        raise ValueError("expected deg >= 0")
+    if data.ndim < 1 or data.ndim != ndim:
+        raise TypeError(f"expected 1D or {ndim}D array for z")
+    if data.size == 0:
+        raise TypeError("expected non-empty vector for z")
+    npoints = len(data)
+    for x in coords:
+        if x.ndim != 1 and x.ndim != ndim:
+            raise TypeError(f"expected 1D or {ndim}D vector for x")
+        if x.size == 0:
+            raise TypeError("expected non-empty vector for x")
+        if len(x) != npoints:
+            raise TypeError("expected coords and data to have same length")
+    if w is not None:
+        if w.ndim != 1 and w.ndim != ndim:
+            raise TypeError(f"expected 1D or {ndim}D vector for w")
+        if data.size != w.size:
+            raise TypeError("expected data and w to have same length")
+    if not callable(vandernd_f) and len(vandernd_f) != ndim:
+        raise TypeError(f"expected a callable or a list of {ndim} vander_1d functions")
+
+    for i in range(ndim):
+        coords[i] = coords[i].ravel()
+    data = data.ravel()
+    if w is not None:
+        w = np.ravel(w)
+
+    if not callable(vandernd_f):
+        vandernd_f2 = lambda coords, deg: _vander_nd_flat(vandernd_f, coords, deg)
+
+    # Remove masked values
+    pxmask = np.ma.getmask(data)
+    for cd in coords:
+        pxmask |= np.ma.getmask(cd)
+    pxmask = ~pxmask
+
+    data = data[pxmask].ravel()
+    for i in range(ndim):
+        coords[i] = coords[i][pxmask].ravel()
+
+    mask = None
+    if deg.ndim == 0:
+        deg = deg[()]
+        deg = np.array([deg] * ndim)
+    elif deg.ndim == 1:
+        deg = deg[:ndim]
+    elif deg.ndim == ndim:
+        mask = deg != 0
+        deg = np.array(deg.shape) - 1
+
+
+    # Calculate elements 1, x, y, x*y, x**2, y**2, ...
+    lhs = vandernd_f2(coords, deg)
+    rhs = data
+
+    # Determine the positions within the vandermode matrix
+    # TODO: how does that work with non standard polynomials?
+    idx = np.zeros((lhs.shape[1], ndim), dtype=int)
+    for i in range(ndim):
+        c = np.ones(ndim)
+        c[i] = 2
+        factors = vandernd_f2(c, deg) // 2
+        idx[:, i] = factors[0]
```

Comment:
I am not very happy with this part, and I don't really know how this will work with not "polynomial" polynomials (e.g. hermite polynomials).